### PR TITLE
feat: start/stop relay ondemand with devserver

### PR DESCRIPTION
**Pull request by untitaker**
Modify the nginx config such that it does not attempt to eagerly resolve hostnames and can be started before relay will be started.

Also revert a commit from #18362 that attempted to simplify the devserver logic, but actually made it impossible to start arbitrary devservices as part of devserver.

With this commit I can finally run nginx/relay/symbolicator/snuba ondemand.

```
for service in SENTRY_DEVSERVICES:
    SENTRY_DEVSERVICES[service]['with_devserver'] = True

# This is necessary for relay tests too and generally useful
SENTRY_DEVSERVICES['redis']['with_devserver'] = False

# XXX: Need to deselect this explicitly? is only_if ignored?
SENTRY_DEVSERVICES['bigtable']['with_devserver'] = False
SENTRY_DEVSERVICES['memcached']['with_devserver'] = False
```

this makes my computer beachball when starting up devserver, but it works

### CODE REVIEW

**Comment by joshuarli** [comment]
@untitaker I set with_devserver to true for relay in my conf, then `sentry devservices up` did not start relay as expected. Here's what my devserver looks like:

```
21:50:19 system        | reverse_proxy started (pid=37700)
21:50:19 system        | webpack started (pid=37698)
21:50:19 system        | relay started (pid=37701)
21:50:19 system        | ingest started (pid=37702)
21:50:19 system        | server started (pid=37703)
-- snip --
21:50:33 relay         | ! Skipping bigtable due to only_if condition
21:50:33 relay         | ! Skipping memcached due to only_if condition
21:50:33 relay         | > Pulling image 'us.gcr.io/sentryio/relay:latest'
21:50:33 reverse_proxy | ! Skipping bigtable due to only_if condition
21:50:33 reverse_proxy | ! Skipping memcached due to only_if condition
-- snip --
21:50:34 reverse_proxy | > Stopping container 'sentry_reverse_proxy'
21:50:34 reverse_proxy | > Removing container 'sentry_reverse_proxy'
21:50:34 reverse_proxy | > Creating container 'sentry_reverse_proxy'
-- snip --
21:50:34 server        | spawned uWSGI master process (pid: 37703)
21:50:34 server        | spawned uWSGI worker 1 (pid: 37736, cores: 2)
21:50:34 reverse_proxy | > Starting container 'sentry_reverse_proxy' (listening: ('127.0.0.1', 8000))
-- snip --
21:51:33 relay         | > Stopping container 'sentry_relay'
21:51:33 relay         | > Removing container 'sentry_relay'
21:51:33 relay         | > Creating container 'sentry_relay'
21:51:33 relay         | > Starting container 'sentry_relay' (listening: ('127.0.0.1', 3000))
21:51:34 relay         | 2020-04-23T21:51:34Z [relay::setup] INFO: launching relay from config folder /etc/relay
```

It works, but it's not without some crinkles. Since relay is pull True, it's pulled everytime devserver is fired up now, which can be quite slow, and I also feel that containers that need to be pulled should be done so only with devservices up. But that's currently the case, I bet people out there have not rerun devservices up in a while and would be running stale services... am I making sense here?

Also only_if is working.

**Comment by jan-auer** [comment]
Can we double check that running sentry tests still work with kafka? 

Somewhat unrelated: I'm realizing that Relay's own tests require Kafka, which means we have to start it manually now.

**Comment by untitaker** [comment]
@jan-auer those are currently permanent

@joshuarli pls check if what I pushed makes sense. `sentry devservices attach` now always starts the service regardless of `with_devserver`, and it also has the `--fast` option which now both skips pulling and recreating the container

**Comment by untitaker** [comment]
@jan-auer btw I am leaning in favor of treating relay and snuba devs as specialcase and making them set with_devserver back to False

**Comment by jan-auer** [comment]
Especially SNUBA seems to consume considerable CPU when idling, so it would be great to have those with the devserver. Also, Relay occupies port 3000, which is a quite common port that might block other projects. 

What's your thinking around having them permanent?

**Comment by untitaker** [comment]
You can opt in and out any of those in your own sentry.conf.py anyway.
Let's defer that discussion.

File: src/sentry/runner/commands/devservices.py [hunk 1]
```diff
@@ -75,9 +75,9 @@ def attach(project, is_devserver, service):
     configure()
 
     client = get_docker_client()
-    containers = _prepare_containers(project)
+    containers = _prepare_containers(project, silent=True)
```
Comment (joshuarli) [inline src/sentry/runner/commands/devservices.py:78]: nice

**Comment by joshuarli** [comment]
relay, reverse_proxy in `sentry devserver` are recreating still (which takes a bit of time):

```
19:59:43 relay         | > Stopping container 'sentry_relay'
19:59:43 relay         | > Removing container 'sentry_relay'
19:59:43 relay         | > Creating container 'sentry_relay'
19:59:43 relay         | > Starting container 'sentry_relay' (listening: ('127.0.0.1', 3000))
```

@untitaker can you review my patch to fix that: 3bf7fac5d19c4555ea2b397fb00fefee3374dc3c

- getting slightly concerned about all the logic spaghetti now lol
- also note that I set `with_devserver` to true for relay, so it's bound to devserver by default for everyone

**Comment by joshuarli** [comment]
Fun note: `devservices up --fast` is so fast now that if you `devserver` ASAP, you might encounter:

```
sentry.exceptions.InvalidConfiguration: Redis is loading the dataset in memory
```

**Comment by untitaker** [comment]
BTW it turns out you practically can't make more than relay and nginx truly ondemand without running into complications when running tests.

I would merge this as-is though because it fixes some bugs for me

**Comment by untitaker** [comment]
@joshuarli oh yeah pls push b98a82a7f1edcd4619658fd5646d96b08f472765 into this branch

**Review by joshuarli — APPROVED** [review]

**Pull request merged**